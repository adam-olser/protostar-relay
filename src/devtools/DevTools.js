/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow
 */

// Reach styles need to come before any component styles.
// This makes overridding the styles simpler.
import '@reach/menu-button/styles.css';
import '@reach/tooltip/styles.css';

import React, { useState, useCallback, useEffect } from 'react';
import type { FrontendBridge } from 'src/bridge';
import Store from './store';
import { BridgeContext, StoreContext } from './context';

import StoreDisplayer from './view/StoreDisplayer';
import NetworkDisplayer from './view/NetworkDisplayer';
import StoreTimeline from './view/StoreTimeline';
// import ComponentsRenderer from './ComponentsRenderer';
// import Network from './Network/Network';
// import StoreInspector from './StoreInspector/StoreInspector';
// import TabBar from './TabBar';
// import { SettingsContextController } from './Settings/SettingsContext';
// import { ModalDialogContextController } from './ModalDialog';



// export type TabID = 'network' | 'settings' | 'store-inspector';
export type ViewElementSource = (id: number) => void;

export type Props = {|
  bridge: FrontendBridge,
    // defaultTab?: TabID,
    // showTabBar?: boolean,
    store: Store,
      viewElementSourceFunction ?: ? ViewElementSource,
      viewElementSourceRequiresFileLocation ?: boolean,

      // This property is used only by the web extension target.
      // The built-in tab UI is hidden in that case, in favor of the browser's own panel tabs.
      // This is done to save space within the app.
      // Because of this, the extension needs to be able to change which tab is active/rendered.
      // overrideTab?: TabID,

      // TODO: Cleanup multi-tabs in webextensions
      // To avoid potential multi-root trickiness, the web extension uses portals to render tabs.
      // The root <DevTools> app is rendered in the top-level extension window,
      // but individual tabs (e.g. Components, Profiling) can be rendered into portals within their browser panels.
      rootContainer ?: Element,
      // networkPortalContainer?: Element,
      settingsPortalContainer ?: Element,
      storeInspectorPortalContainer ?: Element,
|};

const networkTab = {
  id: ('network': TabID),
  icon: 'network',
  label: 'Network',
  title: 'Relay Network',
};
const storeInspectorTab = {
  id: ('store-inspector': TabID),
  icon: 'store-inspector',
  label: 'Store',
  title: 'Relay Store',
};

const tabs = [networkTab, storeInspectorTab];

export default function DevTools({
  bridge,
  // defaultTab = 'store-inspector',
  rootContainer,
  networkPortalContainer,
  storeInspectorPortalContainer,
  // overrideTab,
  settingsPortalContainer,
  // showTabBar = false,
  store,
  viewElementSourceFunction,
  viewElementSourceRequiresFileLocation = false,
}: Props) {
  // const [tab, setTab] = useState(defaultTab);
  // if (overrideTab != null && overrideTab !== tab) {
  //   setTab(overrideTab);
  // }

  const [environmentIDs, setEnvironmentIDs] = useState(
    store.getEnvironmentIDs()
  );
  const [currentEnvID, setCurrentEnvID] = useState(environmentIDs[0]);

  const [selector, setSelector] = useState("Store");

  const setEnv = useCallback(() => {
    const ids = store.getEnvironmentIDs();

    if (currentEnvID === undefined) {
      const firstKey = ids[0];
      setCurrentEnvID(firstKey);
    }
    setEnvironmentIDs(ids);
  }, [store, currentEnvID]);

  useEffect(() => {
    setEnv();
    const testRefresh = () => {
      console.log('hi marc... shuttin \'er down')
    }
    store.addListener('environmentInitialized', setEnv);
    store.addListener('shutdown', testRefresh);
    return () => {
      store.removeListener('environmentInitialized', setEnv);
      store.removeListener('shutdown', testRefresh);
    };
  }, [store, setEnv]);

  function handleTabClick(e, tab) {
    setSelector(tab);
  }

  const handleChange = useCallback(e => {
    setCurrentEnvID(parseInt(e.target.value));
  }, []);

  return (
    <BridgeContext.Provider value={bridge}>
      <StoreContext.Provider value={store}>
        <div className="navigation">
          <form className="env-select">
            <select className="env-select" onChange={handleChange}>
              {environmentIDs.map(id => {
                return (
                  <option key={id} value={id}>{store.getEnvironmentName(id) || id}</option>
                );
              })}
            </select>
          </form>
          <div className="tabs is-toggle">
            <ul>
              <li className={selector === "Store" && "is-active"}>
                <a id="storeSelector" onClick={(e) => handleTabClick(e, "Store")}>
                  <span className="icon is-small">
                    <i className="fas fa-database"></i>
                  </span>
                  <span>Store</span>
                </a>
              </li>
              <li className={selector === "Network" && "is-active"}>
                <a id="networkSelector"
                  onClick={(e) => {
                    handleTabClick(e, "Network");
                  }}
                >
                  <span className="icon is-small">
                    <i className="fas fa-network-wired"></i>
                  </span>
                  <span>Network</span>
                </a>
              </li>
            </ul>
          </div>
        </div>
        <div className={selector === "Store" ? "columns" : "is-hidden"}>
          <StoreTimeline currentEnvID={currentEnvID} portalContainer={storeInspectorPortalContainer} />
        </div>
        <div className={selector === "Network" ? "columns" : "is-hidden"}>
          <NetworkDisplayer />
        </div>
      </StoreContext.Provider>
    </BridgeContext.Provider >
  );
}
