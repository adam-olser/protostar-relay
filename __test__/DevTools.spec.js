import React from 'react';
import { configure, shallow, render } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';
import DevTools from '../src/devtools/DevTools';
import Store from '../src/devtools/store';

configure({ adapter: new Adapter() });

describe('DevTools', () => {
  let wrapper;
  const props = {
    store: {
      getEnvironmentIDs: () => [1],
      getEnvironmentName: (id) => id
    }
  }

  beforeAll(() => {
    wrapper = shallow(<DevTools {...props} />);
  });

  it("My Test Case", () => {
    expect(true).toEqual(true);
  });

});