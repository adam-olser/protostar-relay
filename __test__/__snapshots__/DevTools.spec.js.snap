// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DevTools Renders correctly 1`] = `
Array [
  <div
    className="navigation"
  >
    <form
      className="env-select"
    >
      <select
        className="env-select"
        onChange={[Function]}
      >
        <option
          value={1}
        >
          first
        </option>
        <option
          value={2}
        >
          second
        </option>
        <option
          value={3}
        >
          third
        </option>
      </select>
    </form>
    <div
      className="tabs is-toggle"
    >
      <ul>
        <li
          className="is-active"
        >
          <a
            id="storeSelector"
            onClick={[Function]}
          >
            <span
              className="icon is-small"
            >
              <i
                className="fas fa-database"
              />
            </span>
            <span>
              Store
            </span>
          </a>
        </li>
        <li
          className={false}
        >
          <a
            id="networkSelector"
            onClick={[Function]}
          >
            <span
              className="icon is-small"
            >
              <i
                className="fas fa-network-wired"
              />
            </span>
            <span>
              Network
            </span>
          </a>
        </li>
      </ul>
    </div>
  </div>,
  <div
    className="columns"
  >
    <div
      className="column"
    >
      <div
        className="display-box"
      >
        <div
          className="snapshot-wrapper is-flex"
        >
          <input
            className="input is-small snapshot-btn is-primary"
            onChange={[Function]}
            placeholder="take a store snapshot"
            type="text"
            value=""
          />
          <button
            className="button is-small is-link"
            onClick={[Function]}
          >
            Snapshot
          </button>
        </div>
      </div>
      <div
        className="snapshots"
      >
        <h2
          className="slider-textcolor"
        >
          Store Timeline
        </h2>
        <div
          aria-disabled={false}
          className="input-range"
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onTouchStart={[Function]}
        >
          <span
            className="input-range__label input-range__label--min"
          >
            <span
              className="input-range__label-container"
            >
              0
            </span>
          </span>
          <div
            className="input-range__track input-range__track--background"
            onMouseDown={[Function]}
            onTouchStart={[Function]}
          >
            <div
              className="input-range__track input-range__track--active"
              style={
                Object {
                  "left": "0%",
                  "width": "0%",
                }
              }
            />
            <span
              className="input-range__slider-container"
              style={
                Object {
                  "left": "0%",
                  "position": "absolute",
                }
              }
            >
              <span
                className="input-range__label input-range__label--value"
              >
                <span
                  className="input-range__label-container"
                >
                  0
                </span>
              </span>
              <div
                aria-valuemax={1}
                aria-valuemin={0}
                aria-valuenow={0}
                className="input-range__slider"
                draggable="false"
                onKeyDown={[Function]}
                onMouseDown={[Function]}
                onTouchStart={[Function]}
                role="slider"
                tabIndex="0"
              />
            </span>
          </div>
          <span
            className="input-range__label input-range__label--max"
          >
            <span
              className="input-range__label-container"
            >
              1
            </span>
          </span>
        </div>
        <div
          className="snapshot-nav"
        >
          <button
            class="button is-small"
            onClick={[Function]}
          >
            Backward
          </button>
          <button
            class="button is-small"
            onClick={[Function]}
          >
            Current
          </button>
          <button
            class="button is-small"
            onClick={[Function]}
          >
            Forward
          </button>
        </div>
        <div
          className="snapshot-info"
        />
      </div>
    </div>
    <div
      className="column"
    >
      <p
        class="control has-icons-left"
      >
        <input
          className="input is-small is-primary"
          onChange={[Function]}
          placeholder="Search"
          type="text"
        />
        <button
          className="button is-small is-link"
          onClick={[Function]}
        >
          Reset
        </button>
        <span
          class="icon is-left"
        >
          <i
            class="fas fa-search"
          />
        </span>
      </p>
      <aside
        className="menu"
      >
        <p
          className="menu-label"
        >
          Record List
        </p>
        <ul
          className="menu-list"
        />
      </aside>
    </div>
    <div
      className="column"
    >
      <div
        className="display-box"
      >
        <div
          className="records"
        />
      </div>
    </div>
  </div>,
  <div
    className="is-hidden"
  >
    <div
      className="column is-one-third"
    >
      <p
        class="control has-icons-left"
      >
        <input
          className="input is-small is-primary"
          onChange={[Function]}
          placeholder="Search"
          type="text"
        />
        <button
          className="button is-small is-link"
          onClick={[Function]}
        >
          Reset
        </button>
        <span
          class="icon is-left"
        >
          <i
            class="fas fa-search"
          />
        </span>
      </p>
      <aside
        className="menu"
      >
        <p
          className="menu-label"
        >
          Event List
        </p>
        <ul
          className="menu-list"
        />
      </aside>
    </div>
    <div
      className="column"
    >
      <div
        className="display-box"
      />
    </div>
  </div>,
]
`;
